pinot-common/src/main/java/org/apache/pinot/common/function/scalar/comparison/PolymorphicComparisonScalarFunction.java

```diff
@@ -0,0 +1,75 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.common.function.scalar.comparison;
+
+import javax.annotation.Nullable;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SameOperandTypeChecker;
+import org.apache.pinot.common.function.FunctionInfo;
+import org.apache.pinot.common.function.PinotScalarFunction;
+import org.apache.pinot.common.function.sql.PinotSqlFunction;
+import org.apache.pinot.common.utils.DataSchema;
+
+
+/**
+ * Base class for polymorphic comparison scalar functions
+ */
+public abstract class PolymorphicComparisonScalarFunction implements PinotScalarFunction {
+
+  protected static final double DOUBLE_COMPARISON_TOLERANCE = 1e-7d;
```

Comment:
Are you suggesting using the `=` and `!=` operators directly instead of doing a "fuzzy" comparison?

> We can keep the existing behavior in `getFunctionInfo(int numArguments)` for backward compatibility

Currently, we're simply returning the `double` based comparison function in this case for backward compatibility, so we'll need to create a new "fuzzy" equals / notEquals implementation and return that instead? Also I thought we're planning to move all usages of `getFunctionInfo(int numArguments)` to `getFunctionInfo(ColumnDataType[] argumentTypes)`

Reply:
Yeah. We need this for now so that single-stage engine keeps existing behavior. Can you also check if standard SQL uses exact match or fuzzy match?

Reply:
Looks like it is exact match which is why it isn't recommended to compare float / double values directly. Users should either use some delta value explicitly themselves during comparison or else use the decimal / numeric types for exact precision comparison. I've pushed a commit with the suggested changes.